src/sage/homology/homology_vector_space_with_basis.py

```diff
@@ -841,6 +1208,188 @@ def Sq(self, i):
                 ret += P._from_dict(result, remove_zeros=False)
             return ret
 
+        def _acted_upon_(self, a, self_on_left):
+            r"""
+            Define multiplication of ``self`` by ``a``, an
+            element of the Steenrod algebra.
+
+            INPUT:
+
+            - ``a`` - an element of the mod 2 Steenrod algebra
+            - ``self_on_left`` -- ``True`` if we are computing ``self * a``,
+              otherwise ``a * self``
+
+            Algorithm: for left multiplication by ``a``, since we have
+            :meth:`Sq` to compute multiplication by a single generator
+            `Sq^i`, first convert ``a`` to the Serre-Cartan basis ---
+            that is, sums of products of the elements `Sq^i` --- and
+            then apply :meth:`Sq` repeatedly. Right multiplication by
+            ``a`` is the same as left multiplication by the antipode
+            applied to ``a``.
+
+            EXAMPLES::
+
+                sage: # needs sage.groups
+                sage: SRP4 = simplicial_sets.RealProjectiveSpace(4).suspension()
+                sage: H = SRP4.cohomology_ring(GF(2))
+                sage: x = H.basis()[2,0]
+                sage: Sq(0,1) * x
+                h^{5,0}
+                sage: Sq(3) * x
+                0
+                sage: x * Sq(3)
+                h^{5,0}
+
+            TESTS::
+
+                sage: # needs sage.groups
+                sage: (Sq(2) * Sq(1)) * x == Sq(2) * (Sq(1) * x)
+                True
+                sage: x * (Sq(1) * Sq(2)) == (x * Sq(1)) * Sq(2)
+                True
+            """
+            # Handle field elements first.
+            if a in self.base_ring():
```

Comment:
Same comment about going to CFMElement.

Reply:
Ah, this is the place where I get the infinite recursion. I think I need to keep this one as is.

Reply:
You should be able to do
```python
ret = CombinatorialFreeModule.Element._acted_upon_(self, a, self_on_left)
if ret is not None:  # did the scalar action
    return ret
```
which will be a lot faster than the current implementation. If this doing this does get an infinite loop, can you post the traceback? It really shouldn't have such a loop...